les/fetcher.go

```diff
@@ -618,28 +723,78 @@ func (f *lightFetcher) checkAnnouncedHeaders(fp *fetcherPeerInfo, headers []*typ
 // checkSyncedHeaders updates peer's block tree after synchronisation by marking
 // downloaded headers as known. If none of the announced headers are found after
 // syncing, the peer is dropped.
-func (f *lightFetcher) checkSyncedHeaders(p *peer) {
+func (f *lightFetcher) checkSyncedHeaders(p *peer) (bool, []*types.Header, *big.Int) {
 	fp := f.peers[p]
 	if fp == nil {
 		p.Log().Debug("Unknown peer to check sync headers")
-		return
+		return false, nil, nil
 	}
+
+	var h *types.Header
+	if f.pm.isULCEnabled() {
+		var unapprovedHashes []common.Hash
+		// Overwrite last announced for ULC mode
+		h, unapprovedHashes = f.lastValidTrieNode(p)
+		//rollback untrusted blocks
+		f.chain.Rollback(unapprovedHashes)
+	}
+
 	n := fp.lastAnnounced
 	var td *big.Int
+	trustedHeaderExisted := false
+
+	//find last trusted block
 	for n != nil {
-		if td = f.chain.GetTd(n.hash, n.number); td != nil {
+		//we found last trusted header
+		if n.hash == h.Hash() {
+			trustedHeaderExisted = true
+		}
+		if td = f.chain.GetTd(n.hash, n.number); td != nil && trustedHeaderExisted {
+			break
+		}
+
+		//break if we found last trusted hash before sync
+		if n.hash == f.getLastValidHeader().Hash() {
 			break
 		}
 		n = n.parent
 	}
-	// now n is the latest downloaded header after syncing
-	if n == nil {
+
+	// Now n is the latest downloaded/approved header after syncing
+	if n == nil && !p.isTrusted {
 		p.Log().Debug("Synchronisation failed")
 		go f.pm.removePeer(p.id)
-	} else {
-		header := f.chain.GetHeader(n.hash, n.number)
-		f.newHeaders([]*types.Header{header}, []*big.Int{td})
+		return false, nil, nil
+	}
+	header := f.chain.GetHeader(n.hash, n.number)
+	return true, []*types.Header{header}, td
+}
+
+// lastValidTrieNode return last approved treeNode and a list of unapproved hashes
+func (f *lightFetcher) lastValidTrieNode(p *peer) (*types.Header, []common.Hash) {
+	unapprovedHashes := make([]common.Hash, 0)
+
+	current := f.chain.CurrentHeader()
+	for current != nil || current.Hash() != f.getLastValidHeader().Hash() {
+		if f.isTrustedHash(current.Hash()) {
+			break
+		}
+		unapprovedHashes = append(unapprovedHashes, current.Hash())
+		num := current.Number
+		num.Add(num, big.NewInt(-1))
+		current = f.chain.GetHeader(current.ParentHash, num.Uint64())
 	}
+	return current, unapprovedHashes
+}
+
+func (f *lightFetcher) setLastValidHeader(h *types.Header) {
```

Comment:
This setter function is not really needed since it is only called from a single place but if you need it and want to keep it, call it setLastTrustedHeader instead. Valid and trusted are not the same and both have a meaning in the context of the fetcher.